File: src/commands/system_overview.rs
//! System-wide audit log overview.
//!
//! Provides high-level statistics and insights about Vault usage
//! across the entire audit log. Supports analyzing multiple log files
//! (compressed or uncompressed) for long-term trend analysis.
//!
//! # Usage
//!
//! ```bash
//! # Single file (plain or compressed)
//! vault-audit system-overview audit.log
//! vault-audit system-overview audit.log.gz
//!
//! # Multiple files for week-long analysis
//! vault-audit system-overview day1.log day2.log day3.log
//!
//! # Using shell globbing with compressed files
//! vault-audit system-overview logs/vault_audit.2025-10-*.log.gz
//! ```
//!
//! **Compressed File Support**: Automatically detects and decompresses `.gz` (gzip)
//! and `.zst` (zstandard) files with streaming processing - no temp files needed.
//!
//! # Output
//!
//! Displays comprehensive statistics:
//! - Total entries processed
//! - Unique entities
//! - Unique paths accessed
//! - Operation breakdown (read, write, list, delete)
//! - Top paths by access count
//! - Mount point usage
//! - Authentication method breakdown
//! - Time range covered
//! - Error rate
//!
//! Useful for:
//! - Understanding overall Vault usage
//! - Capacity planning
//! - Identifying hotspots
//! - Security audits

use crate::audit::types::AuditEntry;
use crate::utils::format::format_number;
use crate::utils::parallel::process_files_parallel;
use crate::utils::progress::ProgressBar;
use crate::utils::reader::open_file;
use anyhow::Result;
use std::collections::{HashMap, HashSet};
use std::io::{BufRead, BufReader};

/// Path access statistics
#[derive(Debug)]
struct PathData {
    count: usize,
    operations: HashMap<String, usize>,
    entities: HashSet<String>,
}

impl PathData {
    fn new() -> Self {
        Self {
            count: 0,
            operations: HashMap::with_capacity(10), // Typical: few operation types per path
            entities: HashSet::with_capacity(50),   // Typical: dozens of entities per popular path
        }
    }
}

/// Results from processing a single file
#[derive(Debug)]
struct FileAnalysisResult {
    path_operations: HashMap<String, PathData>,
    operation_types: HashMap<String, usize>,
    path_prefixes: HashMap<String, usize>,
    entity_paths: HashMap<String, HashMap<String, usize>>,
    entity_names: HashMap<String, String>,
}

/// Process entries from a single file using streaming to reduce memory usage
fn process_file_entries_streaming(file_path: &str) -> Result<FileAnalysisResult> {
    use crate::utils::reader::open_file;
    use std::io::{BufRead, BufReader};

    let mut path_operations: HashMap<String, PathData> = HashMap::with_capacity(5000);
    let mut operation_types: HashMap<String, usize> = HashMap::with_capacity(20);
    let mut path_prefixes: HashMap<String, usize> = HashMap::with_capacity(100);
    let mut entity_paths: HashMap<String, HashMap<String, usize>> = HashMap::with_capacity(2000);
    let mut entity_names: HashMap<String, String> = HashMap::with_capacity(2000);

    let file = open_file(file_path)?;
    let reader = BufReader::new(file);

    for line_result in reader.lines() {
        let line = line_result?;

        // Skip empty lines
        if line.trim().is_empty() {
            continue;
        }

        // Parse JSON entry
        let entry: AuditEntry = match serde_json::from_str(&line) {
            Ok(entry) => entry,
            Err(_) => continue, // Skip invalid JSON lines
        };

        let Some(request) = &entry.request else {
            continue;
        };

        let path = match &request.path {
            Some(p) => p.as_str(),
            None => continue,
        };

        let operation = match &request.operation {
            Some(o) => o.as_str(),
            None => continue,
        };

        let entity_id = entry
            .auth
            .as_ref()
            .and_then(|a| a.entity_id.as_deref())
            .unwrap_or("no-entity");

        let display_name = entry
            .auth
            .as_ref()
            .and_then(|a| a.display_name.as_deref())
            .unwrap_or("N/A");

        if path.is_empty() || operation.is_empty() {
            continue;
        }

        // Track by full path
        let path_data = path_operations
            .entry(path.to_string())
            .or_insert_with(PathData::new);
        path_data.count += 1;
        *path_data
            .operations
            .entry(operation.to_string())
            .or_insert(0) += 1;
        path_data.entities.insert(entity_id.to_string());

        // Track by operation type
        *operation_types.entry(operation.to_string()).or_insert(0) += 1;

        // Track by path prefix
        let parts: Vec<&str> = path.trim_matches('/').split('/').collect();
        let prefix = if parts.len() >= 2 {
            format!("{}/{}", parts[0], parts[1])
        } else if !parts.is_empty() {
            parts[0].to_string()
        } else {
            "root".to_string()
        };
        *path_prefixes.entry(prefix).or_insert(0) += 1;

        // Track entity usage
        let entity_map = entity_paths.entry(entity_id.to_string()).or_default();
        *entity_map.entry(path.to_string()).or_insert(0) += 1;
        entity_names
            .entry(entity_id.to_string())
            .or_insert_with(|| display_name.to_string());
    }

    Ok(FileAnalysisResult {
        path_operations,
        operation_types,
        path_prefixes,
        entity_paths,
        entity_names,
    })
}

/// Process entries from a single file (original non-streaming version)
fn process_file_entries(_file_path: &str, entries: Vec<AuditEntry>) -> FileAnalysisResult {
    let mut path_operations: HashMap<String, PathData> = HashMap::with_capacity(5000);
    let mut operation_types: HashMap<String, usize> = HashMap::with_capacity(20);
    let mut path_prefixes: HashMap<String, usize> = HashMap::with_capacity(100);
    let mut entity_paths: HashMap<String, HashMap<String, usize>> = HashMap::with_capacity(2000);
    let mut entity_names: HashMap<String, String> = HashMap::with_capacity(2000);

    for entry in entries {
        let Some(request) = &entry.request else {
            continue;
        };

        let path = match &request.path {
            Some(p) => p.as_str(),
            None => continue,
        };

        let operation = match &request.operation {
            Some(o) => o.as_str(),
            None => continue,
        };

        let entity_id = entry
            .auth
            .as_ref()
            .and_then(|a| a.entity_id.as_deref())
            .unwrap_or("no-entity");

        let display_name = entry
            .auth
            .as_ref()
            .and_then(|a| a.display_name.as_deref())
            .unwrap_or("N/A");

        if path.is_empty() || operation.is_empty() {
            continue;
        }

        // Track by full path
        let path_data = path_operations
            .entry(path.to_string())
            .or_insert_with(PathData::new);
        path_data.count += 1;
        *path_data
            .operations
            .entry(operation.to_string())
            .or_insert(0) += 1;
        path_data.entities.insert(entity_id.to_string());

        // Track by operation type
        *operation_types.entry(operation.to_string()).or_insert(0) += 1;

        // Track by path prefix
        let parts: Vec<&str> = path.trim_matches('/').split('/').collect();
        let prefix = if parts.len() >= 2 {
            format!("{}/{}", parts[0], parts[1])
        } else if !parts.is_empty() {
            parts[0].to_string()
        } else {
            "root".to_string()
        };
        *path_prefixes.entry(prefix).or_insert(0) += 1;

        // Track entity usage
        let entity_map = entity_paths.entry(entity_id.to_string()).or_default();
        *entity_map.entry(path.to_string()).or_insert(0) += 1;
        entity_names
            .entry(entity_id.to_string())
            .or_insert_with(|| display_name.to_string());
    }

    FileAnalysisResult {
        path_operations,
        operation_types,
        path_prefixes,
        entity_paths,
        entity_names,
    }
}

/// Combine results from multiple files
fn combine_results(
    results: Vec<crate::utils::parallel::FileProcessResult<FileAnalysisResult>>,
) -> FileAnalysisResult {
    let mut combined = FileAnalysisResult {
        path_operations: HashMap::with_capacity(5000),
        operation_types: HashMap::with_capacity(20),
        path_prefixes: HashMap::with_capacity(100),
        entity_paths: HashMap::with_capacity(2000),
        entity_names: HashMap::with_capacity(2000),
    };

    for file_result in results {
        let result = file_result.data;

        // Merge path operations
        for (path, path_data) in result.path_operations {
            if let Some(existing) = combined.path_operations.get_mut(&path) {
                existing.count += path_data.count;
                for (op, count) in path_data.operations {
                    *existing.operations.entry(op).or_insert(0) += count;
                }
                existing.entities.extend(path_data.entities);
            } else {
                combined.path_operations.insert(path, path_data);
            }
        }

        // Merge operation types
        for (op, count) in result.operation_types {
            *combined.operation_types.entry(op).or_insert(0) += count;
        }

        // Merge path prefixes
        for (prefix, count) in result.path_prefixes {
            *combined.path_prefixes.entry(prefix).or_insert(0) += count;
        }

        // Merge entity paths
        for (entity_id, paths) in result.entity_paths {
            let entity_map = combined.entity_paths.entry(entity_id).or_default();
            for (path, count) in paths {
                *entity_map.entry(path).or_insert(0) += count;
            }
        }

        // Merge entity names (first occurrence wins)
        for (entity_id, name) in result.entity_names {
            combined.entity_names.entry(entity_id).or_insert(name);
        }
    }

    combined
}

/// Sequential processing fallback for compatibility and single files
fn run_sequential(log_files: &[String]) -> Result<(FileAnalysisResult, usize)> {
    let mut combined_result = FileAnalysisResult {
        path_operations: HashMap::with_capacity(5000),
        operation_types: HashMap::with_capacity(20),
        path_prefixes: HashMap::with_capacity(100),
        entity_paths: HashMap::with_capacity(2000),
        entity_names: HashMap::with_capacity(2000),
    };
    let mut total_lines = 0;

    // Process each file sequentially
    for (file_idx, log_file) in log_files.iter().enumerate() {
        eprintln!(
            "[{}/{}] Processing: {}",
            file_idx + 1,
            log_files.len(),
            log_file
        );

        let file_size = std::fs::metadata(log_file).ok().map(|m| m.len() as usize);
        let mut progress = if let Some(size) = file_size {
            ProgressBar::new(size, "Processing")
        } else {
            ProgressBar::new_spinner("Processing")
        };

        let file = open_file(log_file)?;
        let reader = BufReader::new(file);

        let mut file_lines = 0;
        let mut bytes_read = 0;
        let mut entries = Vec::new();

        for line in reader.lines() {
            file_lines += 1;
            total_lines += 1;
            let line = line?;
            bytes_read += line.len() + 1;

            if file_lines % 10_000 == 0 {
                if let Some(size) = file_size {
                    progress.update(bytes_read.min(size));
                } else {
                    progress.update(file_lines);
                }
            }

            if let Ok(entry) = serde_json::from_str::<AuditEntry>(&line) {
                entries.push(entry);
            }
        }

        if let Some(size) = file_size {
            progress.update(size);
        }

        progress.finish_with_message(&format!(
            "Processed {} lines from this file",
            format_number(file_lines)
        ));

        // Process the entries from this file and merge directly
        let file_result = process_file_entries(log_file, entries);

        // Merge path operations
        for (path, path_data) in file_result.path_operations {
            if let Some(existing) = combined_result.path_operations.get_mut(&path) {
                existing.count += path_data.count;
                for (op, count) in path_data.operations {
                    *existing.operations.entry(op).or_insert(0) += count;
                }
                existing.entities.extend(path_data.entities);
            } else {
                combined_result.path_operations.insert(path, path_data);
            }
        }

        // Merge operation types
        for (op, count) in file_result.operation_types {
            *combined_result.operation_types.entry(op).or_insert(0) += count;
        }

        // Merge path prefixes
        for (prefix, count) in file_result.path_prefixes {
            *combined_result.path_prefixes.entry(prefix).or_insert(0) += count;
        }

        // Merge entity paths
        for (entity_id, paths) in file_result.entity_paths {
            let entity_map = combined_result.entity_paths.entry(entity_id).or_default();
            for (path, count) in paths {
                *entity_map.entry(path).or_insert(0) += count;
            }
        }

        // Merge entity names
        for (entity_id, name) in file_result.entity_names {
            combined_result
                .entity_names
                .entry(entity_id)
                .or_insert(name);
        }
    }

    Ok((combined_result, total_lines))
}

pub fn run(
    log_files: &[String],
    top: usize,
    min_operations: usize,
    sequential: bool,
) -> Result<()> {
    let (combined_result, total_lines) = if sequential || log_files.len() == 1 {
        // Use sequential processing for single files or when explicitly requested
        eprintln!("Processing {} files sequentially...", log_files.len());
        run_sequential(log_files)?
    } else {
        // Use parallel processing for multiple files with streaming
        process_files_parallel(log_files, process_file_entries_streaming, combine_results)?
    };

    eprintln!("\nTotal: Processed {} lines", format_number(total_lines));

    let path_operations = combined_result.path_operations;
    let operation_types = combined_result.operation_types;
    let path_prefixes = combined_result.path_prefixes;
    let entity_paths = combined_result.entity_paths;
    let entity_names = combined_result.entity_names;

    let total_operations: usize = operation_types.values().sum();

    // Print results
    println!("\n{}", "=".repeat(100));
    println!("High-Volume Vault Operations Analysis");
    println!("{}", "=".repeat(100));

    // 1. Operation Types Summary
    println!("\n1. Operation Types (Overall)");
    println!("{}", "-".repeat(100));
    println!("{:<20} {:>15} {:>12}", "Operation", "Count", "Percentage");
    println!("{}", "-".repeat(100));

    let mut sorted_ops: Vec<_> = operation_types.iter().collect();
    sorted_ops.sort_by(|a, b| b.1.cmp(a.1));

    for (op, count) in sorted_ops {
        let pct = if total_operations > 0 {
            (*count as f64 / total_operations as f64) * 100.0
        } else {
            0.0
        };
        println!("{:<20} {:>15} {:>11.2}%", op, format_number(*count), pct);
    }

    println!("{}", "-".repeat(100));
    println!(
        "{:<20} {:>15} {:>11.2}%",
        "TOTAL",
        format_number(total_operations),
        100.0
    );

    // 2. Top Path Prefixes
    println!("\n2. Top Path Prefixes (First 2 components)");
    println!("{}", "-".repeat(100));
    println!(
        "{:<40} {:>15} {:>12}",
        "Path Prefix", "Operations", "Percentage"
    );
    println!("{}", "-".repeat(100));

    let mut sorted_prefixes: Vec<_> = path_prefixes.iter().collect();
    sorted_prefixes.sort_by(|a, b| b.1.cmp(a.1));

    for (prefix, count) in sorted_prefixes.iter().take(top) {
        let pct = if total_operations > 0 {
            (**count as f64 / total_operations as f64) * 100.0
        } else {
            0.0
        };
        println!(
            "{:<40} {:>15} {:>11.2}%",
            prefix,
            format_number(**count),
            pct
        );
    }

    // 3. Top Individual Paths
    println!("\n3. Top {} Individual Paths (Highest Volume)", top);
    println!("{}", "-".repeat(100));
    println!(
        "{:<60} {:>10} {:>10} {:>15}",
        "Path", "Ops", "Entities", "Top Op"
    );
    println!("{}", "-".repeat(100));

    let mut sorted_paths: Vec<_> = path_operations.iter().collect();
    sorted_paths.sort_by(|a, b| b.1.count.cmp(&a.1.count));

    for (path, data) in sorted_paths.iter().take(top) {
        if data.count < min_operations {
            break;
        }
        let top_op = data
            .operations
            .iter()
            .max_by_key(|x| x.1)
            .map_or("N/A", |x| x.0.as_str());
        let path_display = if path.len() > 60 {
            format!("{}...", &path[..58])
        } else {
            (*path).to_string()
        };
        println!(
            "{:<60} {:>10} {:>10} {:>15}",
            path_display,
            format_number(data.count),
            format_number(data.entities.len()),
            top_op
        );
    }

    // 4. Top Entities by Total Operations
    println!("\n4. Top {} Entities by Total Operations", top);
    println!("{}", "-".repeat(100));
    println!(
        "{:<50} {:<38} {:>10}",
        "Display Name", "Entity ID", "Total Ops"
    );
    println!("{}", "-".repeat(100));

    let mut entity_totals: HashMap<String, usize> = HashMap::with_capacity(entity_paths.len());
    for (entity_id, paths) in &entity_paths {
        let total: usize = paths.values().sum();
        entity_totals.insert(entity_id.clone(), total);
    }

    let mut sorted_entities: Vec<_> = entity_totals.iter().collect();
    sorted_entities.sort_by(|a, b| b.1.cmp(a.1));

    for (entity_id, total) in sorted_entities.iter().take(top) {
        let name = entity_names
            .get(*entity_id)
            .map_or("N/A", std::string::String::as_str);
        let name_display = if name.len() > 48 { &name[..48] } else { name };
        let entity_short = if entity_id.len() > 36 {
            &entity_id[..36]
        } else {
            entity_id
        };
        println!(
            "{:<50} {:<38} {:>10}",
            name_display,
            entity_short,
            format_number(**total)
        );
    }

    // 5. Potential Stress Points
    println!("\n5. Potential System Stress Points");
    println!("{}", "-".repeat(100));

    #[derive(Debug)]
    struct StressPoint {
        path: String,
        entity_name: String,
        operations: usize,
    }

    let mut stress_points = Vec::new();

    for (path, data) in &path_operations {
        if data.count >= min_operations {
            for entity_id in &data.entities {
                if let Some(entity_ops_map) = entity_paths.get(entity_id) {
                    if let Some(&entity_ops) = entity_ops_map.get(path) {
                        if entity_ops >= min_operations {
                            stress_points.push(StressPoint {
                                path: path.clone(),
                                entity_name: entity_names
                                    .get(entity_id)
                                    .cloned()
                                    .unwrap_or_else(|| "N/A".to_string()),
                                operations: entity_ops,
                            });
                        }
                    }
                }
            }
        }
    }

    stress_points.sort_by(|a, b| b.operations.cmp(&a.operations));

    println!("{:<40} {:<40} {:>10}", "Entity", "Path", "Ops");
    println!("{}", "-".repeat(100));

    for sp in stress_points.iter().take(top) {
        let entity_display = if sp.entity_name.len() > 38 {
            &sp.entity_name[..38]
        } else {
            &sp.entity_name
        };
        let path_display = if sp.path.len() > 38 {
            &sp.path[..38]
        } else {
            &sp.path
        };
        println!(
            "{:<40} {:<40} {:>10}",
            entity_display,
            path_display,
            format_number(sp.operations)
        );
    }

    println!("{}", "=".repeat(100));
    println!("\nTotal Lines Processed: {}", format_number(total_lines));
    println!("Total Operations: {}", format_number(total_operations));
    println!("{}", "=".repeat(100));

    Ok(())
}

File: internal/commands/system_overview.go
// Package commands provides CLI command implementations for audit analysis.
package commands

import (
	"fmt"
	"sort"

	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/audit"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/processor"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/utils"
)

// PathData holds statistics for a single path.
type PathData struct {
	Count       int
	Operations  map[string]int
	Entities    map[string]bool
}

// SystemOverviewState accumulates statistics across audit entries.
type SystemOverviewState struct {
	PathOperations map[string]*PathData
	OperationTypes map[string]int
	PathPrefixes   map[string]int
	EntityPaths    map[string]map[string]int
	EntityNames    map[string]string
}

// SystemOverviewRun processes audit logs and outputs system overview statistics.
// namespaceFilter, when non-empty, restricts analysis to entries with that namespace ID.
// sequential forces one-file-at-a-time processing even when multiple files are given.
func SystemOverviewRun(logFiles []string, top int, minOps int, namespaceFilter string, sequential bool) error {
	newState := func() SystemOverviewState {
		return SystemOverviewState{
			PathOperations: make(map[string]*PathData),
			OperationTypes: make(map[string]int),
			PathPrefixes:   make(map[string]int),
			EntityPaths:    make(map[string]map[string]int),
			EntityNames:    make(map[string]string),
		}
	}

	cfg := processor.DefaultConfig()
	if sequential {
		cfg.Mode = processor.ModeSequential
	}

	process := func(entry *audit.AuditEntry, state *SystemOverviewState) {
		// Apply namespace filter.
		if namespaceFilter != "" && entry.NamespaceID() != namespaceFilter {
			return
		}

		path := entry.Path()
		operation := entry.Operation()
		entityID := entry.EntityID()
		if entityID == "" {
			entityID = "no-entity"
		}

		if path == "" || operation == "" {
			return
		}

		// Track by full path
		if _, ok := state.PathOperations[path]; !ok {
			state.PathOperations[path] = &PathData{
				Count:      0,
				Operations: make(map[string]int),
				Entities:   make(map[string]bool),
			}
		}
		pathData := state.PathOperations[path]
		pathData.Count++
		pathData.Operations[operation]++
		pathData.Entities[entityID] = true

		// Track by operation type
		state.OperationTypes[operation]++

		// Track by path prefix
		parts := splitPath(path)
		prefix := ""
		if len(parts) >= 2 {
			prefix = parts[0] + "/" + parts[1]
		} else if len(parts) > 0 {
			prefix = parts[0]
		} else {
			prefix = "root"
		}
		state.PathPrefixes[prefix]++

		// Track entity usage
		if _, ok := state.EntityPaths[entityID]; !ok {
			state.EntityPaths[entityID] = make(map[string]int)
		}
		state.EntityPaths[entityID][path]++

		// Store entity display name
		displayName := entry.DisplayName()
		if displayName == "" {
			displayName = "N/A"
		}
		if _, exists := state.EntityNames[entityID]; !exists {
			state.EntityNames[entityID] = displayName
		}
	}

	merge := func(a, b SystemOverviewState) SystemOverviewState {
		// Merge PathOperations
		for path, bData := range b.PathOperations {
			if aData, ok := a.PathOperations[path]; ok {
				aData.Count += bData.Count
				for op, count := range bData.Operations {
					aData.Operations[op] += count
				}
				for entity := range bData.Entities {
					aData.Entities[entity] = true
				}
			} else {
				a.PathOperations[path] = bData
			}
		}

		// Merge OperationTypes
		for op, count := range b.OperationTypes {
			a.OperationTypes[op] += count
		}

		// Merge PathPrefixes
		for prefix, count := range b.PathPrefixes {
			a.PathPrefixes[prefix] += count
		}

		// Merge EntityPaths
		for entity, paths := range b.EntityPaths {
			if _, ok := a.EntityPaths[entity]; !ok {
				a.EntityPaths[entity] = make(map[string]int)
			}
			for path, count := range paths {
				a.EntityPaths[entity][path] += count
			}
		}

		// Merge EntityNames (first one wins)
		for entity, name := range b.EntityNames {
			if _, ok := a.EntityNames[entity]; !ok {
				a.EntityNames[entity] = name
			}
		}

		return a
	}

	result, stats, err := processor.RunFiles(cfg, logFiles, newState, process, merge)
	if err != nil {
		return fmt.Errorf("process files: %w", err)
	}

	stats.Report()

	// Compute total operations
	totalOps := 0
	for _, count := range result.OperationTypes {
		totalOps += count
	}

	// Print results
	fmt.Println("\n" + stringRepeat("=", 100))
	fmt.Println("High-Volume Vault Operations Analysis")
	fmt.Println(stringRepeat("=", 100))

	// 1. Operation Types Summary
	fmt.Println("\n1. Operation Types (Overall)")
	fmt.Println(stringRepeat("-", 100))
	fmt.Printf("%-20s %15s %12s\n", "Operation", "Count", "Percentage")
	fmt.Println(stringRepeat("-", 100))

	type opPair struct {
		name  string
		count int
	}
	var sortedOps []opPair
	for op, count := range result.OperationTypes {
		sortedOps = append(sortedOps, opPair{op, count})
	}
	sort.Slice(sortedOps, func(i, j int) bool {
		return sortedOps[i].count > sortedOps[j].count
	})

	for _, op := range sortedOps {
		pct := float64(0.0)
		if totalOps > 0 {
			pct = float64(op.count) / float64(totalOps) * 100.0
		}
		fmt.Printf("%-20s %15s %11.2f%%\n", op.name, utils.FormatNumber(op.count), pct)
	}

	fmt.Println(stringRepeat("-", 100))
	fmt.Printf("%-20s %15s %11.2f%%\n", "TOTAL", utils.FormatNumber(totalOps), 100.0)

	// 2. Top Path Prefixes
	fmt.Println("\n2. Top Path Prefixes (First 2 components)")
	fmt.Println(stringRepeat("-", 100))
	fmt.Printf("%-40s %15s %12s\n", "Path Prefix", "Operations", "Percentage")
	fmt.Println(stringRepeat("-", 100))

	type prefixPair struct {
		prefix string
		count  int
	}
	var sortedPrefixes []prefixPair
	for prefix, count := range result.PathPrefixes {
		sortedPrefixes = append(sortedPrefixes, prefixPair{prefix, count})
	}
	sort.Slice(sortedPrefixes, func(i, j int) bool {
		return sortedPrefixes[i].count > sortedPrefixes[j].count
	})

	for i, p := range sortedPrefixes {
		if i >= top {
			break
		}
		pct := float64(0.0)
		if totalOps > 0 {
			pct = float64(p.count) / float64(totalOps) * 100.0
		}
		fmt.Printf("%-40s %15s %11.2f%%\n", p.prefix, utils.FormatNumber(p.count), pct)
	}

	// 3. Top Individual Paths
	fmt.Printf("\n3. Top %d Individual Paths (Highest Volume)\n", top)
	fmt.Println(stringRepeat("-", 100))
	fmt.Printf("%-60s %10s %10s %15s\n", "Path", "Ops", "Entities", "Top Op")
	fmt.Println(stringRepeat("-", 100))

	type pathPair struct {
		path string
		data *PathData
	}
	var sortedPaths []pathPair
	for path, data := range result.PathOperations {
		sortedPaths = append(sortedPaths, pathPair{path, data})
	}
	sort.Slice(sortedPaths, func(i, j int) bool {
		return sortedPaths[i].data.Count > sortedPaths[j].data.Count
	})

	for i, p := range sortedPaths {
		if i >= top || p.data.Count < minOps {
			break
		}
		topOp := "N/A"
		maxCount := 0
		for op, count := range p.data.Operations {
			if count > maxCount {
				maxCount = count
				topOp = op
			}
		}
		pathDisplay := p.path
		if len(pathDisplay) > 60 {
			pathDisplay = pathDisplay[:58] + ".."
		}
		fmt.Printf("%-60s %10s %10s %15s\n", pathDisplay, utils.FormatNumber(p.data.Count),
			utils.FormatNumber(len(p.data.Entities)), topOp)
	}

	// 4. Top Entities by Total Operations
	fmt.Printf("\n4. Top %d Entities by Total Operations\n", top)
	fmt.Println(stringRepeat("-", 100))
	fmt.Printf("%-50s %-38s %10s\n", "Display Name", "Entity ID", "Total Ops")
	fmt.Println(stringRepeat("-", 100))

	entityTotals := make(map[string]int)
	for entity, paths := range result.EntityPaths {
		total := 0
		for _, count := range paths {
			total += count
		}
		entityTotals[entity] = total
	}

	type entityPair struct {
		id    string
		total int
	}
	var sortedEntities []entityPair
	for id, total := range entityTotals {
		sortedEntities = append(sortedEntities, entityPair{id, total})
	}
	sort.Slice(sortedEntities, func(i, j int) bool {
		return sortedEntities[i].total > sortedEntities[j].total
	})

	for i, e := range sortedEntities {
		if i >= top {
			break
		}
		name := result.EntityNames[e.id]
		if name == "" {
			name = "N/A"
		}
		nameDisplay := name
		if len(nameDisplay) > 48 {
			nameDisplay = nameDisplay[:48]
		}
		entityShort := e.id
		if len(entityShort) > 36 {
			entityShort = entityShort[:36]
		}
		fmt.Printf("%-50s %-38s %10s\n", nameDisplay, entityShort, utils.FormatNumber(e.total))
	}

	// 5. Potential Stress Points
	fmt.Println("\n5. Potential System Stress Points")
	fmt.Println(stringRepeat("-", 100))

	type stressPoint struct {
		path       string
		entityName string
		operations int
	}
	var stressPoints []stressPoint

	for path, data := range result.PathOperations {
		if data.Count >= minOps {
			for entity := range data.Entities {
				if entityPaths, ok := result.EntityPaths[entity]; ok {
					if entityOps, ok := entityPaths[path]; ok && entityOps >= minOps {
						entityName := result.EntityNames[entity]
						if entityName == "" {
							entityName = "N/A"
						}
						stressPoints = append(stressPoints, stressPoint{path, entityName, entityOps})
					}
				}
			}
		}
	}

	sort.Slice(stressPoints, func(i, j int) bool {
		return stressPoints[i].operations > stressPoints[j].operations
	})

	fmt.Printf("%-40s %-40s %10s\n", "Entity", "Path", "Ops")
	fmt.Println(stringRepeat("-", 100))

	for i, sp := range stressPoints {
		if i >= top {
			break
		}
		entityDisplay := sp.entityName
		if len(entityDisplay) > 38 {
			entityDisplay = entityDisplay[:38]
		}
		pathDisplay := sp.path
		if len(pathDisplay) > 38 {
			pathDisplay = pathDisplay[:38]
		}
		fmt.Printf("%-40s %-40s %10s\n", entityDisplay, pathDisplay, utils.FormatNumber(sp.operations))
	}

	fmt.Println(stringRepeat("=", 100))
	fmt.Printf("\nTotal Lines Processed: %s\n", utils.FormatNumber(stats.TotalLines))
	fmt.Printf("Total Operations: %s\n", utils.FormatNumber(totalOps))
	fmt.Println(stringRepeat("=", 100))

	return nil
}

// splitPath splits a path by "/" and returns components
func splitPath(path string) []string {
	// Trim leading/trailing slashes
	for len(path) > 0 && path[0] == '/' {
		path = path[1:]
	}
	for len(path) > 0 && path[len(path)-1] == '/' {
		path = path[:len(path)-1]
	}

	if path == "" {
		return []string{}
	}

	var result []string
	var current string
	for _, ch := range path {
		if ch == '/' {
			if current != "" {
				result = append(result, current)
				current = ""
			}
		} else {
			current += string(ch)
		}
	}
	if current != "" {
		result = append(result, current)
	}
	return result
}

// stringRepeat repeats a string n times
func stringRepeat(s string, n int) string {
	result := ""
	for i := 0; i < n; i++ {
		result += s
	}
	return result
}
